### PR TITLE
HS-1473 Enrich ValueTest with missing test cases

### PR DESCRIPTION
**Pull request by FrancescoBuc86**
https://opennms.atlassian.net/browse/HS-1473

### CODE REVIEW

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 1]
```diff
@@ -28,30 +28,119 @@
...
+            .writeBytes(new byte[] { 23, 42 })
+            .writeBytes(new byte[] { 13, 37 })
+            .writeBytes(new byte[] { 89, 80 });
+
+        // When
+        final ListValue listValue = (ListValue) ListValue.parserWithBasicList("name1", Optional.empty()).parse(null, byteBuf);
+
+        // Then
+        Assert.assertNotNull(listValue.getValue());
+        Assert.assertEquals(3, listValue.getValue().size());
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:81]: Should we also test for the values here?

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 2]
```diff
@@ -28,30 +28,119 @@
...
+    }
+
+    @Test
+    public void testSubTemplateList() throws MissingTemplateException, InvalidPacketException, UnsupportedEncodingException {
+        // Given
+        final List<Field> fields = List.of(field("field1", "field1".getBytes("UTF-8").length), field("field2", "field2".getBytes("UTF-8").length));
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:87]: Should move `"field1".getBytes("UTF-8")` to a constant as the same value is used as real value below.
Also should use a different length for the second field.
Lastly, should no repeat the field name as value but use something different.

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 3]
```diff
@@ -28,30 +28,119 @@
...
+        Session.Resolver resolver = Mockito.mock(Session.Resolver.class);
+        final Template template = Template.builder(100, Template.Type.OPTIONS_TEMPLATE).withFields(fields).build();
+        Mockito.when(resolver.lookupTemplate(100)).thenReturn(template);
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:90]: Is method order relevant? If not, should move the `when()` calls directly below the `mock()` call. Same goes for the fields and the template (maybe even inline?).

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 4]
```diff
@@ -28,30 +28,119 @@
...
+            .writeBytes("field1".getBytes("UTF-8"))  // Field 1
+            .writeBytes("field2".getBytes("UTF-8"));  // Field 2
+
+        // When
+        final ListValue listValue = (ListValue) ListValue.parserWithSubTemplateList("name1", Optional.empty()).parse(resolver, byteBuf);
+
+        // Then
+        Assert.assertNotNull(listValue.getValue());
+        Assert.assertEquals(1, listValue.getValue().size());
+        Assert.assertEquals(2, listValue.getValue().get(0).size());
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:104]: Check for values?

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 5]
```diff
@@ -28,30 +28,119 @@
...
+            .writeShort(100)  // Template ID
+            .writeBytes("field1".getBytes("UTF-8"))  // Field 1
+            .writeBytes("field2".getBytes("UTF-8"));  // Field 2
+
+        // When
+        final ListValue listValue = (ListValue) ListValue.parserWithSubTemplateList("name1", Optional.empty()).parse(resolver, byteBuf);
+
+        // Then
+        Assert.assertNotNull(listValue.getValue());
+        Assert.assertEquals(1, listValue.getValue().size());
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:103]: Add a second instance to the data?

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 6]
```diff
@@ -28,30 +28,119 @@
...
+    private static Field field(String name, final int length) {
+        return new Field() {
+            @Override
+            public int length() {
+                    return length;
+            }
+
+            @Override
+            public Value<?> parse(Session.Resolver resolver, ByteBuf buffer) {
+                return new OctetArrayValue(name, bytes(buffer, length()));
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:116]:
```suggestion
                return new OctetArrayValue(name, bytes(buffer, length));
```

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 7]
```diff
@@ -28,30 +28,119 @@
...
+            .writeShort(16) // Header length
+            .writeBytes("field1".getBytes("UTF-8"))  // Field 1
+            .writeBytes("field2".getBytes("UTF-8"));  // Field 2
+
+        // When
+        final ListValue listValue = (ListValue) ListValue.parserWithSubTemplateMultiList("name1", Optional.empty()).parse(resolver, byteBuf);
+
+        // Then
+        Assert.assertNotNull(listValue.getValue());
+        Assert.assertEquals(2, listValue.getValue().get(0).size());
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:142]: Same comments as for `testSubTemplateList` apply here, too.

**Review by fooker — CHANGES_REQUESTED** [review]

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 1 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:81]: sure

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 2 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:87]: ok

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 3 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:90]: ok

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 4 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:104]: sure

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 5 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java]: sure

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 7 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java]: ok

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 1 above)
Comment (fooker) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:82]: To improvements:
- You can directly cast to the value returned by `getValue`. Should be pretty stable.
- Split up the values we expect to something like `23 * 0x100 + 42` which would explain why these match.

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 8]
```diff
@@ -28,30 +28,155 @@
...
+
+    @Test
+    public void testSubTemplateList() throws MissingTemplateException, InvalidPacketException {
+        // Given
+        String field1Name = "firstField";
+        String field1Value = "firstFieldValue";
+        String field2Name = "secondField";
+        String field2Value = "secondFieldValue";
+        byte[] field1 = field1Value.getBytes(StandardCharsets.UTF_8);
+        byte[] field2 = field2Value.getBytes(StandardCharsets.UTF_8);
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:96]:
```suggestion
        byte[] field2Name = "secondField".getBytes();
        byte[] field2Value = "secondFieldValue".getBytes();
```

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 9]
```diff
@@ -28,30 +28,155 @@
...
+        final List<Field> fields = List.of(field(field1Name, field1.length), field(field2Name, field2.length));
+        Session.Resolver resolver = Mockito.mock(Session.Resolver.class);
+        Mockito.when(resolver.lookupTemplate(100)).thenReturn(Template.builder(100, Template.Type.OPTIONS_TEMPLATE)
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:99]: Should be a `Template.Type.TEMPLATE`

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java [hunk 10]
```diff
@@ -28,30 +28,155 @@
...
+            .withFields(fields).build());
+
+        ByteBuf byteBuf = Unpooled.buffer()
+            .writeByte(0xFF) // Semantic: Undefined
+            .writeShort(357) // Header ID
+            .writeShort(dataRecordsLength + field1.length + field2.length + field1.length + field2.length)  // Header length
+            .writeBytes(field1)  // Field 1 list 1
+            .writeBytes(field2)  // Field 2 list 1
+            .writeBytes(field1)  // Field 1 list 2
+            .writeBytes(field2); // Field 2 list 2
```
Comment (fooker) [inline minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:158]: We should add a second template with some more data here.

**Review by fooker — CHANGES_REQUESTED** [review]

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 9 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:99]: ok

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 1 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java]: done

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 10 above)
Comment (fooker) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:162]: Can we add a different set of fields to the second template?

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 1 above)
Comment (fooker) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java]: That's the way the values is build from the values we write in the raw bytes.

File: minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java (hunk 10 above)
Comment (FrancescoBuc86) [reply minion/minion-flows/flows-parser/src/test/java/org/opennms/horizon/minion/flows/parser/ValueTest.java:162]: yep, updated

**Pull request merged**